text/BrunoPitrus/galil.tex

```diff
@@ -0,0 +1,117 @@
+
+W tym opisie, tak samo jak w załączonej implementacji algorytmu, stringi indeksowane są od 1. Wszelki kod należy interpretować zgodnie z semantyką języka Python.
+
+Niech \texttt{m = len(word)}, \texttt{n = len(text)}. Rozważmy schemat algorytmu dopasowania wzorca
+\begin{alltt}
+p,q= 0,0
+\textbf{while} True:
+  \textbf{while} text[p+q+1] == word[q+1]: q+= 1
+  \textbf{if} q == m: \textbf{yield} p+1
+  p,q= \textit{p'},\textit{q'}
+\end{alltt}
+W powyższym pseudokodzie \texttt{p} to jest obecny kandydat na pozycję wzorca, a \texttt{q} to prefiks o którym wiemy że się zgadza.
+
+Naiwny algorytm \textit{brute-force} używa \texttt{p'= p+1}, \texttt{q'= 0}. Ponieważ wiemy jednak że \texttt{text[p+1:p+q+1] == word[1:q+1]} to rozważanie \texttt{p'= p+x} ma sens tylko jeśli \texttt{word[1:q-x+1] == word[x+1:q+1]}; na tej obserwacji jest oparty algorytm Knutha–Morrisa–Pratta, obliczający \texttt{p'= p+shift[q], q'= q-shift[q] if q>0 else 0}, gdzie \texttt{shift[q]} jest długością najkrótszego okresu \texttt{word[1:q+1]}.
+
+Algorytm Galila–Seiferasa jest oparty na tym schemacie oraz następującym twierdzeniu:
+\begin{theorem}[O dekompozycji]
+Dla ustalonego (odpowiednio dużego) \(k\), każde słowo \texttt{x} można przedstawić jako \texttt{u+v}, gdzie \texttt{v} ma co najwyżej jeden \textit{\(k\)-okres prefiksu} a \texttt{u} jest długości \(\symrm O{\left(\text{długość najkrótszego okresu \texttt{v}}\right)}\). 
+\end{theorem}
+Słowo \texttt{z} nazywamy \(k\)-okresem prefiksu słowa \texttt{w} jeśli \texttt{z} jest słowem pierwotnym a \texttt{z\textsuperscript{k}} jest prefiksem \texttt{w}.
+
+Do skonstruowania tej dekompozycji przyda nam się następujący lemat:
+\begin{lemma}
+Dla słowa \textit{pierwotnego} \texttt{w} istnieje dekompozycja \texttt{w == w1 + w2} taka że dla dowolnego słowa \texttt{w'}, słowo \texttt{w2 + w\textsuperscript{k-1} + w'} nie ma \(k\)-okresu prefiksu krótszego niż \texttt{len(w)}.
+\end{lemma}
+
+Pierwsza część algorytmu jest poświęcona znalezieniu dekompozycji wzorca takiej jak w twierdzeniu o dekompozycji.
+\begin{alltt}
+s= 0
+\textbf{while} word[s+1:] \textit{ma więcej niż jeden k-okres prefiksu}:
+  p2= \textit{długość drugiego najkrótszego k-okresu prefiksu}
+  \textit{Korzystając z lematu, znajdź}
+   s' < s + p2 \textit{takie że} word[s'+1:] \textit{nie ma k-okresu prefiksu krótszego niż} p2
+  s= s'
+\end{alltt}
+Niech \texttt{l(s)} będzie długością najkrótszego okresu \texttt{word[s+1:]} a \texttt{p1(s)} długością najkrótszego \(k\)-okresu prefiksu. Indukcyjnie można dowieść, że w każdej iteracji pętli \texttt{s < c * min(l(s), p1(s))} dla \texttt{c = (k-1)/(k-2)} — więc w szczególności na końcu \texttt{len(u) < c * l(s)}.

```

Comment:
Is the inductive proof long and complicated? Do you considered putting its idea here?